examples/camera-app/linux/src/clusters/webrtc-provider/webrtc-provider-manager.cpp

```diff
@@ -887,42 +887,45 @@ void WebRTCProviderManager::OnConnectionStateChanged(bool connected, const uint1
     else
     {
-        WebrtcTransport * transport = GetTransport(sessionId);
-        if (transport == nullptr)
-        {
-
-            ChipLogProgress(Camera,
-                            "Transport not found for session %u during disconnect; session may have already been cleaned up",
-                            sessionId);
-            return;
-        }
+        // Schedule cleanup on Matter thread to ensure proper locking when calling RemoveSession
+        DeviceLayer::SystemLayer().ScheduleLambda([this, sessionId]() {
+            WebrtcTransport * transport = GetTransport(sessionId);
+            if (transport == nullptr)
+            {
+                ChipLogProgress(Camera,
+                                "Transport not found for session %u during disconnect; session may have already been cleaned up",
+                                sessionId);
+                return;
+            }
 
-        // Connection was closed/disconnected by the peer - clean up the session
-        ChipLogProgress(Camera, "Peer connection closed for session %u, cleaning up resources", sessionId);
+            // Connection was closed/disconnected by the peer - clean up the session
+            ChipLogProgress(Camera, "Peer connection closed for session %u, cleaning up resources", sessionId);
 
-        // Release the Video and Audio Streams from the CameraAVStreamManagement
-        // cluster and update the reference counts.
-        ReleaseAudioVideoStreams(sessionId);
+            // Release the Video and Audio Streams from the CameraAVStreamManagement
+            // cluster and update the reference counts.
+            ReleaseAudioVideoStreams(sessionId);
 
-        // Capture args before unregistering in case the transport is invalidated
-        WebrtcTransport::RequestArgs args = transport->GetRequestArgs();
+            // Capture args before unregistering in case the transport is invalidated
+            WebrtcTransport::RequestArgs args = transport->GetRequestArgs();
 
-        // Unregister the transport from the media controller
-        UnregisterWebrtcTransport(sessionId);
+            // Unregister the transport from the media controller
+            UnregisterWebrtcTransport(sessionId);
 
-        // Remove from session maps
-        mSessionIdMap.erase(ScopedNodeId(args.peerNodeId, args.fabricIndex));
+            // Remove from session maps
+            mSessionIdMap.erase(ScopedNodeId(args.peerNodeId, args.fabricIndex));
 
-        // Remove from current sessions list in the WebRTC Transport Provider
-        if (mWebRTCTransportProvider != nullptr)
-        {
-            mWebRTCTransportProvider->RemoveSession(sessionId);
-        }
+            // Remove from current sessions list in the WebRTC Transport Provider
+            // This MUST be called on the Matter thread with the stack lock held
+            if (mWebRTCTransportProvider != nullptr)
+            {
+                mWebRTCTransportProvider->RemoveSession(sessionId);
```

Comment:
RemoveSession() is the one that needs to run on the Matter thread, right? Are we bundling the entire clean-up logic in the lambda to keep the execution together(which also may not be guaranteed to happen)?

Reply:
Before I saw the PR I was assuming that it would just scheduleLamda for the RemoveSession.  But it may be cleaner treating the whole cleanup as atomic, and we're also calling down into the SDK in ReleaseAudioVideoStreams, which is safe for now, but may not be depending on how that code evolves.